Basic auth extension

jon-wei opened the pull request:
This PR adds an extension that provides implementations of the auth interfaces introduced in #4271 :
- an Authenticator which supports HTTP Basic authentication
- an Escalator which supports HTTP Basic authentication
- an Authorizer which implements basic role-based access control

With the large amount of changes (almost a rewrite) from the last round of PR reviews on #4823, I felt like it would be easier to close the old PR and start a fresh one.

Some design notes:
- The user/role information is stored in the metadata store, using the "druid_config" table as a key-value store, user/role maps are stored as serialized byte arrays
- On the coordinator a *MetadataStorageUpdater class is responsible for handling writes to the user/role DB and holds a deserialized cached copy of the latest DB state
- Each node has a *CacheManager instance that polls the coordinator for the latest DB state periodically, this also syncs the auth DB state at startup time
- The coordinator can also notify and push updates to remote nodes, using the *CacheNotifier classes
- The BasicHTTPAuthenticator and BasicRoleBasedAuthorizer perform their checks using cached info from the *CacheManager instance, to avoid overloading the coordinator
- The coordinator nodes have a jersey resource for configuring the auth DB, which also holds the endpoint that nodes poll to get the latest DB state
- A jersey resource is also installed on non-coordinator nodes, containing an endpoint that listens for DB state updates

jihoonson commented:
@jon-wei thanks for raising a new PR, but it looks that this single patch contains quite large changes. Is it possible to split into several smaller PRs? It would be much helpful for reviewers.

jon-wei commented:
@jihoonson

I can split the PR into two if really needed, but the changes are pretty self contained in an extension with few changes to core druid. Within the extension the authentication and authorization portions are also pretty self-contained. I'd rather keep this as one PR if possible since the features are closely related and to avoid the churn/overhead of having two PRs to go through the review process.

gianm commented:
I'm reviewing this now.

jihoonson commented:
@jon-wei ok. I'm also reviewing.

jihoonson commented on extensions-core/druid-basic-security/src/main/java/io/druid/security/basic/authentication/BytesFullResponseHandler.java:
```diff
@@ -0,0 +1,75 @@
+      ClientResponse<FullResponseHolder> clientResponse, Throwable e
+  )
+  {
+    // Its safe to Ignore as the ClientResponse returned in handleChunk were unfinished
```
Would you elaborate on this?

jon-wei replied:
this was a comment copied from `FullResponseHandler`

jihoonson commented on extensions-core/druid-basic-security/src/main/java/io/druid/security/basic/authentication/db/cache/BasicAuthenticatorCacheNotifier.java:
```diff
@@ -0,0 +1,25 @@
+
+package io.druid.security.basic.authentication.db.cache;
+
+public interface BasicAuthenticatorCacheNotifier
```
Please add some javadoc.

jon-wei replied:
Added javadocs

jihoonson commented on extensions-core/druid-basic-security/src/main/java/io/druid/security/basic/authentication/db/cache/CoordinatorBasicAuthenticatorCacheNotifier.java:
```diff
@@ -0,0 +1,280 @@
+                      LOG.debug("Got status: " + srh.getStatus());
+                    }
+                    catch (Exception e) {
+                      LOG.makeAlert("Failed to get response for cache notification.").emit();
```
```e``` should be included in the alert.

jon-wei replied:
Added e to alert

jihoonson commented on extensions-core/druid-basic-security/src/main/java/io/druid/security/basic/authentication/db/cache/CoordinatorBasicAuthenticatorCacheNotifier.java:
```diff
@@ -0,0 +1,280 @@
+      notifierThread = Execs.makeThread(
+          "CoordinatorBasicAuthenticatorCacheNotifier-notifierThread",
+          () -> {
+            while (!Thread.interrupted()) {
```
I wonder why you created a single thread which repeatedly sends update notifications instead of using an ExecutorService and submitting a new task on addUpdate(). I guess this will be simpler.

jon-wei replied:
Hm, there wasn't a particular reason, do you see an advantage to using ExecutorService with new tasks?

jihoonson replied:
I think using ExecutorService and BlockingQueue will be simpler to implement/maintain the code. Please see the comments (https://github.com/druid-io/druid/pull/5099#discussion_r156345477 and https://github.com/druid-io/druid/pull/5099#discussion_r156347623)

jihoonson commented on extensions-core/druid-basic-security/src/main/java/io/druid/security/basic/authentication/db/cache/CoordinatorBasicAuthenticatorCacheNotifier.java:
```diff
@@ -0,0 +1,280 @@
+        ClientResponse<StatusResponseHolder> clientResponse, Throwable e
+    )
+    {
+      // Its safe to Ignore as the ClientResponse returned in handleChunk were unfinished
```
Would you elaborate on this?

jon-wei replied:
This was copied from `FullResponseHandler`

jihoonson commented on extensions-core/druid-basic-security/src/main/java/io/druid/security/basic/authentication/db/cache/NoopBasicAuthenticatorCacheNotifier.java:
```diff
@@ -0,0 +1,29 @@
+
+package io.druid.security.basic.authentication.db.cache;
+
+public class NoopBasicAuthenticatorCacheNotifier implements BasicAuthenticatorCacheNotifier
```
I wonder this is needed.

jon-wei replied:
Deleted this class

jihoonson commented on extensions-core/druid-basic-security/src/main/java/io/druid/security/basic/authentication/endpoint/BasicAuthenticatorResourceHandler.java:
```diff
@@ -0,0 +1,39 @@
+
+import javax.ws.rs.core.Response;
+
+public interface BasicAuthenticatorResourceHandler
```
Would you add some javadoc? It looks that the implementation of some methods should be different depending on the node type (coordinator or others).

jon-wei replied:
Added javadocs

jihoonson commented on extensions-core/druid-basic-security/src/main/java/io/druid/security/basic/authentication/endpoint/CoordinatorBasicAuthenticatorResourceHandler.java:
```diff
@@ -0,0 +1,192 @@
+
+public class CoordinatorBasicAuthenticatorResourceHandler implements BasicAuthenticatorResourceHandler
+{
+  private static final Logger log = new Logger(CoordinatorBasicAuthenticatorResourceHandler.class);
```
Unused variable.

jon-wei replied:
Removed

jihoonson commented on extensions-core/druid-basic-security/src/main/java/io/druid/security/basic/authentication/endpoint/CoordinatorBasicAuthenticatorResourceHandler.java:
```diff
@@ -0,0 +1,192 @@
+  @Override
+  public Response authenticatorUpdateListener(String authenticatorName, byte[] serializedUserMap)
+  {
+    return Response.status(Response.Status.NOT_FOUND).build();
```
Maybe ```FORBIDDEN```.

jon-wei replied:
The coordinator/non-coordinator APIs are combined in one resource, because there wasn't a way for me to selectively create the resource based on node type in the extension module.

So I want to use NOT_FOUND, the idea being that the unsupported APIs should appear to the user as "not existing", rather than "exists but is forbidden"

jihoonson replied:
Thanks. Makes sense.

jihoonson commented on extensions-core/druid-basic-security/src/main/java/io/druid/security/basic/authentication/entity/BasicAuthenticatorCredentials.java:
```diff
@@ -0,0 +1,101 @@
+  )
+  {
+    this.salt = salt;
+    this.hash = hash;
```
Please add ```Preconditions.notNull()``` check for the above two variables.

jon-wei replied:
Add preconditions check

jihoonson commented on extensions-core/druid-basic-security/src/main/java/io/druid/security/basic/authorization/db/cache/BasicAuthorizerCacheManager.java:
```diff
@@ -0,0 +1,34 @@
+
+import java.util.Map;
+
+public interface BasicAuthorizerCacheManager
```
Please add some javadoc. It looks that some methods need to be implemented differently according to the node type (coordinator or others).

jon-wei replied:
Added some javadocs

jihoonson commented on extensions-core/druid-basic-security/src/main/java/io/druid/security/basic/authorization/db/cache/CoordinatorBasicAuthorizerCacheManager.java:
```diff
@@ -0,0 +1,58 @@
+
+import java.util.Map;
+
+public class CoordinatorBasicAuthorizerCacheManager implements BasicAuthorizerCacheManager
```
I think ```MetadataStoragePollingBasicAuthorizerCacheManager``` is more intuitive.

jon-wei replied:
Renamed

jihoonson commented on extensions-core/druid-basic-security/src/main/java/io/druid/security/basic/authorization/db/cache/CoordinatorBasicAuthorizerCacheNotifier.java:
```diff
@@ -0,0 +1,277 @@
+import java.util.concurrent.TimeUnit;
+
+@ManageLifecycle
+public class CoordinatorBasicAuthorizerCacheNotifier implements BasicAuthorizerCacheNotifier
```
Similar comments for ```CoordinatorBasicAuthenticatorCacheNotifier```. Please refactor to share the common code paths.

jon-wei replied:
Refactored

jihoonson commented on extensions-core/druid-basic-security/src/main/java/io/druid/security/basic/authorization/db/cache/DefaultBasicAuthorizerCacheManager.java:
```diff
@@ -0,0 +1,236 @@
+import java.util.concurrent.TimeUnit;
+
+@ManageLifecycle
+public class DefaultBasicAuthorizerCacheManager implements BasicAuthorizerCacheManager
```
I think ```CoordinatorPollingBasicAuthorizerCacheManager``` is more intuitive.

jon-wei replied:
Renamed

jihoonson commented on extensions-core/druid-basic-security/src/main/java/io/druid/security/basic/authorization/db/cache/DefaultBasicAuthorizerCacheManager.java:
```diff
@@ -0,0 +1,236 @@
+  private final DruidLeaderClient druidLeaderClient;
+  private final BasicAuthCommonCacheConfig commonCacheConfig;
+
+  private volatile ScheduledExecutorService exec;
```
Why is this volatile? And it should be shut down on stop().

jon-wei replied:
Fixed this, similar to previous comments

jihoonson commented on extensions-core/druid-basic-security/src/main/java/io/druid/security/basic/authorization/db/cache/DefaultBasicAuthorizerCacheManager.java:
```diff
@@ -0,0 +1,236 @@
+
+              long randomDelay = ThreadLocalRandom.current().nextLong(0, commonCacheConfig.getMaxRandomDelay());
+              LOG.debug("Inserting random polling delay of [%s] ms", randomDelay);
+              Thread.sleep(randomDelay);
```
Better sleep at the beginning because every node will try to sync simultaneously at first if sleep after polling.

jon-wei replied:
Moved sleep

jihoonson commented on extensions-core/druid-basic-security/src/main/java/io/druid/security/basic/authorization/db/cache/DefaultBasicAuthorizerCacheManager.java:
```diff
@@ -0,0 +1,236 @@
+      cachedRoleMaps.put(authorizerPrefix, userAndRoleMap.getRoleMap());
+    }
+    catch (IOException ioe) {
+      LOG.makeAlert("WTF? Could not deserialize user map received from coordinator.").emit();
```
ioe should be included in the alert.

jon-wei replied:
added ioe to alert

jihoonson commented on extensions-core/druid-basic-security/src/main/java/io/druid/security/basic/authorization/db/cache/DefaultBasicAuthorizerCacheManager.java:
```diff
@@ -0,0 +1,236 @@
+      Authorizer authorizer = entry.getValue();
+      if (authorizer instanceof BasicRoleBasedAuthorizer) {
+        String authorizerName = entry.getKey();
+        BasicRoleBasedAuthorizer roleBasedAuthorizer = (BasicRoleBasedAuthorizer) authorizer;
```
Unused variable.

jon-wei replied:
Removed

jihoonson commented on extensions-core/druid-basic-security/src/main/java/io/druid/security/basic/authorization/db/cache/NoopBasicAuthorizerCacheNotifier.java:
```diff
@@ -0,0 +1,29 @@
+
+package io.druid.security.basic.authorization.db.cache;
+
+public class NoopBasicAuthorizerCacheNotifier implements BasicAuthorizerCacheNotifier
```
Is this needed?

jon-wei replied:
It's unneeded, I deleted this class

jihoonson commented on extensions-core/druid-basic-security/src/main/java/io/druid/security/basic/authorization/db/updater/BasicAuthorizerMetadataStorageUpdater.java:
```diff
@@ -0,0 +1,56 @@
+
+  Map<String, BasicAuthorizerUser> deserializeUserMap(byte[] userMapBytes);
+
+  Map<String, BasicAuthorizerRole> deserializeRoleMap(byte[] roleMapBytes);
```
Suggest to move to some util class like ```BasicAuthUtils``` to avoid duplicated codes.

jon-wei replied:
Moved these to BasicAuthUtils

jihoonson commented on extensions-core/druid-basic-security/src/main/java/io/druid/security/basic/authorization/db/updater/CoordinatorBasicAuthorizerMetadataStorageUpdater.java:
```diff
@@ -0,0 +1,761 @@
+
+  public static final List<ResourceAction> SUPERUSER_PERMISSIONS = makeSuperUserPermissions();
+
+  public static final String USERS = "users";
```
Can be private.

jon-wei replied:
Made private

jihoonson commented on extensions-core/druid-basic-security/src/main/java/io/druid/security/basic/authorization/db/updater/CoordinatorBasicAuthorizerMetadataStorageUpdater.java:
```diff
@@ -0,0 +1,761 @@
+  public static final List<ResourceAction> SUPERUSER_PERMISSIONS = makeSuperUserPermissions();
+
+  public static final String USERS = "users";
+  public static final String ROLES = "roles";
```
Can be private.

jihoonson commented on extensions-core/druid-basic-security/src/main/java/io/druid/security/basic/authorization/db/updater/CoordinatorBasicAuthorizerMetadataStorageUpdater.java:
```diff
@@ -0,0 +1,761 @@
+  public static final TypeReference ROLE_MAP_TYPE_REFERENCE = new TypeReference<Map<String, BasicAuthorizerRole>>()
+  {
+  };
+  public static final TypeReference USER_AND_ROLE_MAP_TYPE_REFERENCE = new TypeReference<UserAndRoleMap>()
```
Suggest to move the above TypeReferences to some util class to remove duplicated codes in ```CoordinatorBasicAuthenticatorMetadataStorageUpdater```.

jon-wei replied:
Moved type references to BasicAuthUtil

jihoonson commented on extensions-core/druid-basic-security/src/main/java/io/druid/security/basic/authorization/db/updater/CoordinatorBasicAuthorizerMetadataStorageUpdater.java:
```diff
@@ -0,0 +1,761 @@
+  };
+
+  public static final String DEFAULT_ADMIN_NAME = "admin";
+  public static final String DEFAULT_INTERNAL_SYSTEM_NAME = "druid_system";
```
Can use ```ADMIN_NAME``` and ```INTERNAL_USER_NAME``` of ```BasicAuthUtils```.

jon-wei replied:
Changed this

jihoonson commented on extensions-core/druid-basic-security/src/main/java/io/druid/security/basic/authorization/db/updater/CoordinatorBasicAuthorizerMetadataStorageUpdater.java:
```diff
@@ -0,0 +1,761 @@
+  private final Map<String, byte[]> cachedSerializedUserMaps;
+
+  private final Map<String, Map<String, BasicAuthorizerRole>> cachedRoleMaps;
+  private final Map<String, byte[]> cachedSerializedRoleMaps;
```
Looks that we can delegate thread-safety to ConcurrentHashMap if cachedUserMaps and cachedRoleMaps are changed to something like ConcurrentHashMap<String, Pair<Map<String, BasicAuthorizerUser(/Role)>, byte[]>>. Also suggest to create a class to represent Pair<Map<String, BasicAuthorizerUser(/Role)>, byte[]> for better readability.

jon-wei replied:
Added *Bundle class for the combined map and serialized version, changed to ConcurrentHashMaps

jihoonson commented on extensions-core/druid-basic-security/src/main/java/io/druid/security/basic/authorization/db/updater/CoordinatorBasicAuthorizerMetadataStorageUpdater.java:
```diff
@@ -0,0 +1,761 @@
+  private final Set<String> authorizerNames;
+  private final LifecycleLock lifecycleLock = new LifecycleLock();
+
+  private volatile ScheduledExecutorService exec;
```
Why is this volatile? Also, it should be shut down on stop().

jon-wei replied:
Fixed

jihoonson commented on extensions-core/druid-basic-security/src/main/java/io/druid/security/basic/authorization/db/updater/CoordinatorBasicAuthorizerMetadataStorageUpdater.java:
```diff
@@ -0,0 +1,761 @@
+  private final LifecycleLock lifecycleLock = new LifecycleLock();
+
+  private volatile ScheduledExecutorService exec;
+  private volatile boolean stopped = false;
```
Suggest to add isStopped() to LifecycleLock.

jihoonson commented on extensions-core/druid-basic-security/src/main/java/io/druid/security/basic/authorization/db/updater/CoordinatorBasicAuthorizerMetadataStorageUpdater.java:
```diff
@@ -0,0 +1,761 @@
+  private void createUserInternal(String prefix, String userName)
+  {
+    int attempts = 0;
+    while (attempts < numRetries) {
```
Maybe backoff is needed between retries?

jon-wei replied:
Added backoff

jihoonson commented on extensions-core/druid-basic-security/src/main/java/io/druid/security/basic/authorization/db/updater/NoopBasicAuthorizerMetadataStorageUpdater.java:
```diff
@@ -0,0 +1,110 @@
+import java.util.List;
+import java.util.Map;
+
+public class NoopBasicAuthorizerMetadataStorageUpdater implements BasicAuthorizerMetadataStorageUpdater
```
I think this is not needed.

jon-wei replied:
Removed

jihoonson commented on extensions-core/druid-basic-security/src/main/java/io/druid/security/basic/authorization/endpoint/BasicAuthorizerResourceHandler.java:
```diff
@@ -0,0 +1,54 @@
+import javax.ws.rs.core.Response;
+import java.util.List;
+
+public interface BasicAuthorizerResourceHandler
```
Please add some javadoc. This class also contains some methods to be implemented depending on the node type.

jon-wei replied:
Add javadocs

jihoonson commented on extensions-core/druid-basic-security/src/main/java/io/druid/security/basic/authorization/endpoint/CoordinatorBasicAuthorizerResourceHandler.java:
```diff
@@ -0,0 +1,402 @@
+
+  private final BasicAuthorizerMetadataStorageUpdater storageUpdater;
+  private final Map<String, BasicRoleBasedAuthorizer> authorizerMap;
+  private final ObjectMapper objectMapper;
```
Unused variable

jon-wei replied:
This is used now

jihoonson commented on extensions-core/druid-basic-security/src/main/java/io/druid/security/basic/authorization/endpoint/CoordinatorBasicAuthorizerResourceHandler.java:
```diff
@@ -0,0 +1,402 @@
+  @Override
+  public Response authorizerUpdateListener(String authorizerName, byte[] serializedUserAndRoleMap)
+  {
+    return Response.status(Response.Status.NOT_FOUND).build();
```
Maybe ```FORBIDDEN```.

jihoonson commented on extensions-core/druid-basic-security/src/main/java/io/druid/security/basic/authorization/endpoint/DefaultBasicAuthorizerResourceHandler.java:
```diff
@@ -0,0 +1,159 @@
+public class DefaultBasicAuthorizerResourceHandler implements BasicAuthorizerResourceHandler
+{
+  private static final Logger log = new Logger(DefaultBasicAuthorizerResourceHandler.class);
+  private static final Response NOT_FOUND_RESPONSE = Response.status(Response.Status.NOT_FOUND).build();
```
Maybe ```FORBIDDEN```

jihoonson commented on extensions-core/druid-basic-security/src/main/java/io/druid/security/basic/authorization/entity/BasicAuthorizerPermission.java:
```diff
@@ -0,0 +1,121 @@
+      this.resourceNamePattern = Pattern.compile(resourceAction.getResource().getName());
+    }
+    catch (PatternSyntaxException pse) {
+      throw new BasicSecurityDBResourceException(
```
```pse``` should be included.

jon-wei replied:
Included exception

jihoonson commented on extensions-core/druid-basic-security/src/main/java/io/druid/security/basic/authorization/entity/BasicAuthorizerPermission.java:
```diff
@@ -0,0 +1,121 @@
+
+    BasicAuthorizerPermission that = (BasicAuthorizerPermission) o;
+
+    if (getResourceAction() != null
```
You can use ```Objects.equals()```.

jon-wei replied:
This was autogenerated by IntelliJ

jihoonson commented on extensions-core/druid-basic-security/src/main/java/io/druid/security/basic/authorization/entity/BasicAuthorizerPermission.java:
```diff
@@ -0,0 +1,121 @@
+  public int hashCode()
+  {
+    int result = getResourceAction() != null ? getResourceAction().hashCode() : 0;
+    result = 31 * result + (getResourceNamePattern().pattern() != null
```
You can use ```Objects.hash()```.

jon-wei replied:
Also autogenerated, they're not used outside of tests

jihoonson commented on extensions-core/druid-basic-security/src/test/java/io/druid/security/authentication/CoordinatorBasicAuthenticatorMetadataStorageUpdaterTest.java:
```diff
@@ -0,0 +1,200 @@
+    connector = derbyConnectorRule.getConnector();
+    tablesConfig = derbyConnectorRule.metadataTablesConfigSupplier().get();
+    connector.createConfigTable();
+    //injector = setupInjector();
```
Please remove this line.

jon-wei replied:
removed line

gianm commented:
Thanks @jon-wei for working on the extension and @jihoonson for reviewing.

Beyond what @jihoonson has said (especially an API to check the load status -- it's always helpful) I think we should cache the credential db on the local disk of each node, so when they start up, they can use their cache if the coordinator is not accessible. Maybe they should use their cache if they can't contact the coordinator within some amount of time. That way the coordinator being down doesn't torch the cluster completely.

jon-wei commented on docs/content/development/extensions-core/druid-basic-security.md:
```diff
@@ -0,0 +1,275 @@
+
+Please see [Authentication and Authorization](../../configuration/auth.html) for more information on the extension interfaces being implemented.
+
+## Configuration
```
Added a note about `common.runtime.properties` here

jon-wei commented on docs/content/development/extensions-core/druid-basic-security.md:
```diff
@@ -0,0 +1,275 @@
+#### Properties
+|Property|Description|Default|required|
+|--------|-----------|-------|--------|
+|`druid.auth.authenticator.MyBasicAuthenticator.initialAdminPassword`|Initial password for the automatically created default admin user. If no password is specified, the default admin user will not be created.|null|No|
```
initial is the password these users will have when the user is first automatically created.

The password can be changed afterwards and won't be reset after cluster restart, I added a comment about that.

jon-wei commented on docs/content/development/extensions-core/druid-basic-security.md:
```diff
@@ -0,0 +1,275 @@
+|`druid.auth.authenticator.MyBasicAuthenticator.cacheNotificationTimeout`|The timeout in milliseconds for the cache notifications.|5000|No|
+|`druid.auth.authenticator.MyBasicAuthenticator.authorizerName`|Authorizer that requests should be directed to|N/A|Yes|
+
+enableCacheNotifications
```
deleted, thanks

jon-wei commented on docs/content/development/extensions-core/druid-basic-security.md:
```diff
@@ -0,0 +1,275 @@
+|Property|Description|Default|required|
+|--------|-----------|-------|--------|
+|`druid.escalator.internalClientUsername`|The escalator will use this username for requests made as the internal systerm user.|n/a|Yes|
+|`druid.escalator.internalClientPassword`|The escalator will use this password for requests made as the internal system user.|n/a|Yes|
```
It should be set to the same value, unless the user changes the password later on

jon-wei commented on docs/content/development/extensions-core/druid-basic-security.md:
```diff
@@ -0,0 +1,275 @@
+Each API endpoint includes {authenticatorName}, specifying which Authenticator instance is being configured.
+
+##### User/Credential Management
+`GET(/{authenticatorName}/users)`
```
Added full paths to these.

jon-wei commented on docs/content/development/extensions-core/druid-basic-security.md:
```diff
@@ -0,0 +1,275 @@
+
+`POST(/{authenticatorName}/users/{userName}/credentials)`
+Assign a password used for HTTP basic authentication for {userName}
+Content: password string
```
I changed this API slightly, there is a JSON request body example now

jon-wei commented on docs/content/development/extensions-core/druid-basic-security.md:
```diff
@@ -0,0 +1,275 @@
+Return a list of all user names.
+
+`GET(/{authorizerName}/users/{userName})`
+Return the name and credentials information of the user with name {userName}
```
Updated the comment

jon-wei commented on docs/content/development/extensions-core/druid-basic-security.md:
````diff
@@ -0,0 +1,275 @@
+```
+[
+{ 
+  resource": {
````
FIxed

jon-wei commented on extensions-core/druid-basic-security/src/main/java/io/druid/security/basic/BasicAuthUtils.java:
```diff
@@ -0,0 +1,104 @@
+public class BasicAuthUtils
+{
+  private static final Logger log = new Logger(BasicAuthUtils.class);
+  private static final Base64.Encoder ENCODER = Base64.getEncoder();
```
Got rid of ENCODER and DECODER

jon-wei commented on extensions-core/druid-basic-security/src/main/java/io/druid/security/basic/BasicAuthUtils.java:
```diff
@@ -0,0 +1,104 @@
+  public final static String ADMIN_NAME = "admin";
+  public final static String INTERNAL_USER_NAME = "druid_system";
+
+  public static int SALT_LENGTH = 32;
```
Added some comments on the values

jon-wei commented on extensions-core/druid-basic-security/src/main/java/io/druid/security/basic/BasicAuthUtils.java:
```diff
@@ -0,0 +1,104 @@
+  public final static String INTERNAL_USER_NAME = "druid_system";
+
+  public static int SALT_LENGTH = 32;
+  public static int KEY_ITERATIONS = 10000;
```
I changed the iteration counts so that the default value is a property on each Authenticator, and there's also an option to specify an iteration count on the credential update API. This makes it possible to update the iteration count by resetting the password using the API.

jon-wei commented on extensions-core/druid-basic-security/src/main/java/io/druid/security/basic/BasicAuthUtils.java:
```diff
@@ -0,0 +1,104 @@
+      throw new RuntimeException(ikse);
+    }
+    catch (NoSuchAlgorithmException nsae) {
+      log.error("%s not supported on this system.", ALGORITHM);
```
Fixed

jon-wei commented on extensions-core/druid-basic-security/src/main/java/io/druid/security/basic/BasicAuthUtils.java:
```diff
@@ -0,0 +1,104 @@
+    return salt;
+  }
+
+  public static String getBasicUserSecretFromHttpReq(HttpServletRequest httpReq)
```
Added @Nullable

jon-wei commented on extensions-core/druid-basic-security/src/main/java/io/druid/security/basic/BasicAuthUtils.java:
```diff
@@ -0,0 +1,104 @@
+      return StringUtils.fromUtf8(DECODER.decode(encodedUserSecret));
+    }
+    catch (Exception e) {
+      return null;
```
Moved the length checks out, and catches exceptions only from the Base64 decode now

jon-wei commented on extensions-core/druid-basic-security/src/main/java/io/druid/security/basic/BasicSecurityDBResourceException.java:
```diff
@@ -0,0 +1,34 @@
+import io.druid.java.util.common.StringUtils;
+
+/**
+ * Throw this in the BasicSecurityStorageConnectors for invalid resource accesses that are likely a result of user error
```
Updated this comment

jon-wei commented on extensions-core/druid-basic-security/src/main/java/io/druid/security/basic/BasicSecurityDruidModule.java:
```diff
@@ -0,0 +1,187 @@
+    if (isCoordinator(injector)) {
+      return injector.getInstance(CoordinatorBasicAuthenticatorMetadataStorageUpdater.class);
+    } else {
+      return injector.getInstance(NoopBasicAuthenticatorMetadataStorageUpdater.class);
```
It's not, I have this return null now

jon-wei commented on extensions-core/druid-basic-security/src/main/java/io/druid/security/basic/BasicSecurityResourceFilter.java:
```diff
@@ -0,0 +1,89 @@
+public class BasicSecurityResourceFilter extends AbstractResourceFilter
+{
+  private static final List<String> applicablePaths = ImmutableList.of(
+      "druid/coordinator/v1/security/*",
```
Removed the *

jon-wei commented on extensions-core/druid-basic-security/src/main/java/io/druid/security/basic/BasicSecurityResourceFilter.java:
```diff
@@ -0,0 +1,89 @@
+  public ContainerRequest filter(ContainerRequest request)
+  {
+    final ResourceAction resourceAction = new ResourceAction(
+        new Resource("security", ResourceType.CONFIG),
```
Made this static

jon-wei commented on extensions-core/druid-basic-security/src/main/java/io/druid/security/basic/authentication/db/cache/BasicAuthenticatorCacheManager.java:
```diff
@@ -0,0 +1,31 @@
+
+import java.util.Map;
+
+public interface BasicAuthenticatorCacheManager
```
Added some javadocs

jon-wei commented on extensions-core/druid-basic-security/src/main/java/io/druid/security/basic/authentication/db/cache/CoordinatorBasicAuthenticatorCacheManager.java:
```diff
@@ -0,0 +1,55 @@
+
+import java.util.Map;
+
+public class CoordinatorBasicAuthenticatorCacheManager implements BasicAuthenticatorCacheManager
```
Renamed

jon-wei commented on extensions-core/druid-basic-security/src/main/java/io/druid/security/basic/authentication/db/cache/CoordinatorBasicAuthenticatorCacheNotifier.java:
```diff
@@ -0,0 +1,280 @@
+import java.util.concurrent.TimeUnit;
+
+@ManageLifecycle
+public class CoordinatorBasicAuthenticatorCacheNotifier implements BasicAuthenticatorCacheNotifier
```
Moved the common code into a `CommonCacheNotifier`

jon-wei commented on extensions-core/druid-basic-security/src/main/java/io/druid/security/basic/authentication/db/cache/DefaultBasicAuthenticatorCacheManager.java:
```diff
@@ -0,0 +1,222 @@
+import java.util.concurrent.TimeUnit;
+
+@ManageLifecycle
+public class DefaultBasicAuthenticatorCacheManager implements BasicAuthenticatorCacheManager
```
Renamed

jon-wei commented on extensions-core/druid-basic-security/src/main/java/io/druid/security/basic/authentication/db/cache/DefaultBasicAuthenticatorCacheManager.java:
```diff
@@ -0,0 +1,222 @@
+  private final DruidLeaderClient druidLeaderClient;
+  private final BasicAuthCommonCacheConfig commonCacheConfig;
+
+  private volatile ScheduledExecutorService exec;
```
Moved initialization of this to constructor, removed volatile, and added a stop()

jon-wei commented on extensions-core/druid-basic-security/src/main/java/io/druid/security/basic/authentication/db/cache/DefaultBasicAuthenticatorCacheManager.java:
```diff
@@ -0,0 +1,222 @@
+
+              long randomDelay = ThreadLocalRandom.current().nextLong(0, commonCacheConfig.getMaxRandomDelay());
+              LOG.debug("Inserting random polling delay of [%s] ms", randomDelay);
+              Thread.sleep(randomDelay);
```
Moved the sleep to the beginning

jon-wei commented on extensions-core/druid-basic-security/src/main/java/io/druid/security/basic/authentication/db/cache/DefaultBasicAuthenticatorCacheManager.java:
```diff
@@ -0,0 +1,222 @@
+      );
+    }
+    catch (IOException ioe) {
+      LOG.makeAlert("WTF? Could not deserialize user map received from coordinator.").emit();
```
Added ioe to alert

jon-wei commented on extensions-core/druid-basic-security/src/main/java/io/druid/security/basic/authentication/db/cache/DefaultBasicAuthenticatorCacheManager.java:
```diff
@@ -0,0 +1,222 @@
+      Authenticator authenticator = entry.getValue();
+      if (authenticator instanceof BasicHTTPAuthenticator) {
+        String authenticatorName = entry.getKey();
+        BasicHTTPAuthenticator basicHTTPAuthenticator = (BasicHTTPAuthenticator) authenticator;
```
Removed

jon-wei commented on extensions-core/druid-basic-security/src/main/java/io/druid/security/basic/authentication/db/updater/BasicAuthenticatorMetadataStorageUpdater.java:
```diff
@@ -0,0 +1,43 @@
+
+import java.util.Map;
+
+public interface BasicAuthenticatorMetadataStorageUpdater
```
Added some javadocs

jon-wei commented on extensions-core/druid-basic-security/src/main/java/io/druid/security/basic/authentication/db/updater/BasicAuthenticatorMetadataStorageUpdater.java:
```diff
@@ -0,0 +1,43 @@
+
+  Map<String, BasicAuthenticatorUser> deserializeUserMap(byte[] userMapBytes);
+
+  byte[] serializeUserMap(Map<String, BasicAuthenticatorUser> userMap);
```
Changed these to static methods in BasicAuthUtils

jon-wei commented on extensions-core/druid-basic-security/src/main/java/io/druid/security/basic/authentication/db/updater/CoordinatorBasicAuthenticatorMetadataStorageUpdater.java:
```diff
@@ -0,0 +1,408 @@
+  private static final EmittingLogger LOG =
+      new EmittingLogger(CoordinatorBasicAuthenticatorMetadataStorageUpdater.class);
+
+  public static final String USERS = "users";
```
Made this private

jon-wei commented on extensions-core/druid-basic-security/src/main/java/io/druid/security/basic/authentication/db/updater/CoordinatorBasicAuthenticatorMetadataStorageUpdater.java:
```diff
@@ -0,0 +1,408 @@
+      new EmittingLogger(CoordinatorBasicAuthenticatorMetadataStorageUpdater.class);
+
+  public static final String USERS = "users";
+  public static final TypeReference USER_MAP_TYPE_REFERENCE = new TypeReference<Map<String, BasicAuthenticatorUser>>()
```
Moved the type references to BasicAuthUtils

jon-wei commented on extensions-core/druid-basic-security/src/main/java/io/druid/security/basic/authentication/db/updater/CoordinatorBasicAuthenticatorMetadataStorageUpdater.java:
```diff
@@ -0,0 +1,408 @@
+  private final int numRetries = 5;
+
+  private final Map<String, Map<String, BasicAuthenticatorUser>> cachedUserMaps;
+  private final Map<String, byte[]> cachedSerializedUserMaps;
```
Added a BasicAuthenticatorUserMapBundle class, changed this to ConcurrentHashMap

jon-wei commented on extensions-core/druid-basic-security/src/main/java/io/druid/security/basic/authentication/db/updater/CoordinatorBasicAuthenticatorMetadataStorageUpdater.java:
```diff
@@ -0,0 +1,408 @@
+  private final Set<String> authenticatorPrefixes;
+  private final LifecycleLock lifecycleLock = new LifecycleLock();
+
+  private volatile ScheduledExecutorService exec;
```
Moved exec creation to constructor, removed volatile, added stop()

jon-wei commented on extensions-core/druid-basic-security/src/main/java/io/druid/security/basic/authentication/db/updater/CoordinatorBasicAuthenticatorMetadataStorageUpdater.java:
```diff
@@ -0,0 +1,408 @@
+  private final LifecycleLock lifecycleLock = new LifecycleLock();
+
+  private volatile ScheduledExecutorService exec;
+  private volatile boolean stopped = false;
```
Hm, let's look into that in another PR, since a lot of classes use Lifecycle

jihoonson replied:
Ok.

jon-wei commented on extensions-core/druid-basic-security/src/main/java/io/druid/security/basic/authentication/db/updater/CoordinatorBasicAuthenticatorMetadataStorageUpdater.java:
```diff
@@ -0,0 +1,408 @@
+          cachedUserMaps.put(authenticatorName, userMap);
+          cachedSerializedUserMaps.put(authenticatorName, userMapBytes);
+
+          if (dbConfig.getInitialAdminPassword() != null && !userMap.containsKey("admin")) {
```
Changed this to use ADMIN_NAME

jon-wei commented on extensions-core/druid-basic-security/src/main/java/io/druid/security/basic/authentication/db/updater/CoordinatorBasicAuthenticatorMetadataStorageUpdater.java:
```diff
@@ -0,0 +1,408 @@
+            setUserCredentialsInternal(authenticatorName, "admin", dbConfig.getInitialAdminPassword().toCharArray());
+          }
+
+          if (dbConfig.getInitialInternalClientPassword() != null && !userMap.containsKey("druid_system")) {
```
Changed this to use INTERNAL_USER_NAME

jon-wei commented on extensions-core/druid-basic-security/src/main/java/io/druid/security/basic/authentication/db/updater/CoordinatorBasicAuthenticatorMetadataStorageUpdater.java:
```diff
@@ -0,0 +1,408 @@
+  private void createUserInternal(String prefix, String userName)
+  {
+    int attempts = 0;
+    while (attempts < numRetries) {
```
Added a random backoff between retires

jon-wei commented on extensions-core/druid-basic-security/src/main/java/io/druid/security/basic/authentication/db/updater/NoopBasicAuthenticatorMetadataStorageUpdater.java:
```diff
@@ -0,0 +1,77 @@
+  @Override
+  public void createUser(String prefix, String userName)
+  {
+    throw new UnsupportedOperationException("not supported");
```
Deleted the class

jon-wei commented:
@jihoonson @gianm 

Thanks for the reviews, I addressed the comments and added the following features:
- Disk snapshots of the caches, these will be used during initialization if a service cannot sync with the coordinator
- A "refresh all" cache syncing API on the coordinator
- loadStatus API (I also added "/db/" to some of the API paths to prevent path conflicts")

jihoonson commented on extensions-core/druid-basic-security/src/main/java/io/druid/security/basic/authentication/endpoint/CoordinatorBasicAuthenticatorResourceHandler.java:
```diff
@@ -0,0 +1,217 @@
+  )
+  {
+    this.storageUpdater = storageUpdater;
+    this.objectMapper = new ObjectMapper(new SmileFactory());
```
Looks that objectMapper can be injected using ```@Smile``` annotation.

jon-wei replied:
Changed to use `@Smile` annotation

jihoonson commented on extensions-core/druid-basic-security/src/main/java/io/druid/security/basic/authorization/db/cache/CoordinatorPollingBasicAuthorizerCacheManager.java:
```diff
@@ -0,0 +1,291 @@
+      }
+    }
+    catch (Exception e) {
+      LOG.makeAlert(e, "WTF? Could not deserialize user/role map received from coordinator.").emit();
```
Is this intended to does nothing for all kinds of exceptions but emitting an alert?

jon-wei replied:
Yes, see the previous comment

jihoonson commented on extensions-core/druid-basic-security/src/main/java/io/druid/security/basic/authorization/db/cache/CoordinatorPollingBasicAuthorizerCacheManager.java:
```diff
@@ -0,0 +1,291 @@
+  }
+
+  @Nullable
+  private UserAndRoleMap loadUserAndRoleMapFromDisk(String prefix) throws Exception
```
Can be reduced to ```throws IOException```.

jon-wei replied:
Reduced to IOException

jihoonson commented on extensions-core/druid-basic-security/src/main/java/io/druid/security/basic/authorization/db/cache/CoordinatorPollingBasicAuthorizerCacheManager.java:
```diff
@@ -0,0 +1,291 @@
+    );
+  }
+
+  private void writeMapToDisk(String prefix, byte[] userMapBytes) throws Exception
```
Can be reduced to IOException.

jon-wei replied:
Reduced to IOException

jihoonson commented on extensions-core/druid-basic-security/src/main/java/io/druid/security/basic/authorization/db/cache/CoordinatorPollingBasicAuthorizerCacheManager.java:
```diff
@@ -0,0 +1,291 @@
+    File cacheDir = new File(commonCacheConfig.getCacheDirectory());
+    cacheDir.mkdirs();
+    File userMapFile = new File(commonCacheConfig.getCacheDirectory(), getUserRoleMapFilename(prefix));
+    Files.write(userMapBytes, userMapFile);
```
Similar comment. What if the write fails?

jon-wei commented on extensions-core/druid-basic-security/src/main/java/io/druid/security/basic/BasicAuthUtils.java:
```diff
@@ -0,0 +1,235 @@
+      return objectMapper.writeValueAsBytes(userMap);
+    }
+    catch (IOException ioe) {
+      throw new ISE("WTF? Couldn't serialize userMap!");
```
added

jon-wei commented on extensions-core/druid-basic-security/src/main/java/io/druid/security/basic/BasicAuthUtils.java:
```diff
@@ -0,0 +1,235 @@
+      return objectMapper.writeValueAsBytes(userMap);
+    }
+    catch (IOException ioe) {
+      throw new ISE("WTF? Couldn't serialize userMap!");
```
added

jon-wei commented on extensions-core/druid-basic-security/src/main/java/io/druid/security/basic/BasicAuthUtils.java:
```diff
@@ -0,0 +1,235 @@
+      return objectMapper.writeValueAsBytes(roleMap);
+    }
+    catch (IOException ioe) {
+      throw new ISE("WTF? Couldn't serialize roleMap!");
```
added

jon-wei commented on extensions-core/druid-basic-security/src/main/java/io/druid/security/basic/CommonCacheNotifier.java:
```diff
@@ -0,0 +1,238 @@
+  private final EmittingLogger logger;
+  private final String threadName;
+
+  private Thread notifierThread;
```
Added stop(), changed to use ExecutorService

jon-wei commented on extensions-core/druid-basic-security/src/main/java/io/druid/security/basic/CommonCacheNotifier.java:
```diff
@@ -0,0 +1,238 @@
+      HttpClient httpClient,
+      String baseUrl,
+      String threadName,
+      EmittingLogger logger
```
hm, I feel like it'd be more useful to use this logger passed in from the authenticator/authorization classes, so that the logs appear differently based on the class that's using this common notifier

jihoonson replied:
Ah it's needed. But I'm still concerned with that it's difficult to find where the logging code is from the log if the logging class is different from the actual class. What do you think about passing a caller name in constructor and adding it to all logs?

jon-wei commented on extensions-core/druid-basic-security/src/main/java/io/druid/security/basic/CommonCacheNotifier.java:
```diff
@@ -0,0 +1,238 @@
+              HashMap<String, byte[]> serializedMapsSnapshot;
+              synchronized (itemsToUpdate) {
+                if (itemsToUpdate.isEmpty()) {
+                  itemsToUpdate.wait();
```
changed to BlockingQueue

jon-wei commented on extensions-core/druid-basic-security/src/main/java/io/druid/security/basic/CommonCacheNotifier.java:
```diff
@@ -0,0 +1,238 @@
+                );
+                for (ListenableFuture<StatusResponseHolder> future : futures) {
+                  try {
+                    StatusResponseHolder srh = future.get(
```
Changed to Futures.allAsList

jon-wei commented on extensions-core/druid-basic-security/src/main/java/io/druid/security/basic/CommonCacheNotifier.java:
```diff
@@ -0,0 +1,238 @@
+    synchronized (itemsToUpdate) {
+      itemsToUpdate.add(updatedItemName);
+      serializedMaps.put(updatedItemName, updatedItemData);
+      itemsToUpdate.notify();
```
I think the current behavior is okay for now, since the updates should be pretty infrequent, I changed this to use Pair<String, byte[]> and a BlockingQueue

jon-wei commented on extensions-core/druid-basic-security/src/main/java/io/druid/security/basic/CommonCacheNotifier.java:
```diff
@@ -0,0 +1,238 @@
+        ClientResponse<StatusResponseHolder> clientResponse, Throwable e
+    )
+    {
+      // Its safe to Ignore as the ClientResponse returned in handleChunk were unfinished
```
Added an error log here

jon-wei commented on extensions-core/druid-basic-security/src/main/java/io/druid/security/basic/authentication/db/cache/CoordinatorPollingBasicAuthenticatorCacheManager.java:
```diff
@@ -0,0 +1,280 @@
+      }
+    }
+    catch (Exception e) {
+      LOG.makeAlert(e, "WTF? Could not deserialize user map received from coordinator.").emit();
```
Yes, that was the intent, I don't think there's a whole lot it can do in case of errors except try again

jon-wei commented on extensions-core/druid-basic-security/src/main/java/io/druid/security/basic/authentication/db/cache/CoordinatorPollingBasicAuthenticatorCacheManager.java:
```diff
@@ -0,0 +1,280 @@
+  }
+
+  @Nullable
+  private Map<String, BasicAuthenticatorUser> loadUserMapFromDisk(String prefix) throws Exception
```
Reduced to IOException

jon-wei commented on extensions-core/druid-basic-security/src/main/java/io/druid/security/basic/authentication/db/cache/CoordinatorPollingBasicAuthenticatorCacheManager.java:
```diff
@@ -0,0 +1,280 @@
+    );
+  }
+
+  private void writeUserMapToDisk(String prefix, byte[] userMapBytes) throws Exception
```
Reduced to IOException

jon-wei commented on extensions-core/druid-basic-security/src/main/java/io/druid/security/basic/authentication/db/cache/CoordinatorPollingBasicAuthenticatorCacheManager.java:
```diff
@@ -0,0 +1,280 @@
+    File cacheDir = new File(commonCacheConfig.getCacheDirectory());
+    cacheDir.mkdirs();
+    File userMapFile = new File(commonCacheConfig.getCacheDirectory(), getUserMapFilename(prefix));
+    Files.write(userMapBytes, userMapFile);
```
I think it's fine as-is, the read will do nothing if the file is malformed and can't be deserialized, and the file would eventually get replaced on a successful cache poll/notification

jihoonson commented on extensions-core/druid-basic-security/src/main/java/io/druid/security/basic/CommonCacheNotifier.java:
```diff
@@ -0,0 +1,237 @@
+  private final Map<String, BasicAuthDBConfig> itemConfigMap;
+  private final String baseUrl;
+  private final EmittingLogger logger;
+  private final String threadName;
```
Unused variable.

jihoonson commented on extensions-core/druid-basic-security/src/main/java/io/druid/security/basic/authentication/db/cache/CoordinatorPollingBasicAuthenticatorCacheManager.java:
```diff
@@ -0,0 +1,281 @@
+            return loadUserMapFromDisk(prefix);
+          }
+          catch (Exception e2) {
+            LOG.makeAlert(e2, "Encountered exception while loading user map snapshot for authenticator [%s]", prefix);
```
Please add ```e``` to ```e2``` like ```e2.addSuppressed(e)```.

jihoonson commented on extensions-core/druid-basic-security/src/main/java/io/druid/security/basic/authorization/db/cache/CoordinatorPollingBasicAuthorizerCacheManager.java:
```diff
@@ -0,0 +1,292 @@
+            return loadUserAndRoleMapFromDisk(prefix);
+          }
+          catch (Exception e2) {
+            LOG.makeAlert(e2, "Encountered exception while loading user-role map snapshot for authorizer [%s]", prefix);
```
Please add ```e``` to ```e2```.

jihoonson commented:
@jon-wei thanks for the quick update! +1 after CI.

jon-wei commented:
@jihoonson thanks for the review!

leventov commented on extensions-core/druid-basic-security/src/main/java/io/druid/security/basic/CommonCacheNotifier.java:
```diff
@@ -0,0 +1,236 @@
+{
+  private static final EmittingLogger LOG = new EmittingLogger(CommonCacheNotifier.class);
+
+  private static final List<String> NODE_TYPES = Arrays.asList(
```
@jon-wei does it purposely skip the Coordinator type?

jon-wei replied:
@leventov Yes, the coordinator gets its information about the auth state directly from metadata storage, this update notification is for other service types

zhouyuanchao commented:
Why basic security can't use EnvironmentVariablePasswordProvider, such as initialAdminPassword?

jon-wei commented:
@zhouyuanchao It can, check the master branch

zhouyuanchao commented:
@jon-wei internalClientPassword can not use EnvironmentVariablePasswordProvider in master branch
